### ISSUE TITLE
Huge performance problems with rails >= 4.0

### ISSUE THREAD

**Issue by dalyons** [thiagopradi/octopus#280]
tl;dr - something in the octopus AR metaprogramming is causing huge amounts of constant cache invalidations, which is disastrous for performance. 

So we have a bigish, high traffic rails API site. Recently did the bump from 3.2 to 4.0, and our 95% response times basically doubled, and response times were choppy all over the place. Using New Relic, we were able to see this suspicious graph:

![screen shot 2014-09-30 at 2 46 01 pm](https://cloud.githubusercontent.com/assets/43060/4467108/41d19e10-48ec-11e4-8e63-06b9b5f707d2.png)

After a lot of troubleshooting using dtrace & the pretty cool Busted gem https://github.com/simeonwillbanks/busted , we were able to verify that under rails 4.0 we were invalidating the ruby constant cache sometimes up to hundreds of times per request (as opposed to zero times in rails 3.2) 

Removing ar-octopus and changing nothing else solved the problem and led to this: 

![screen shot 2014-09-30 at 2 39 29 pm](https://cloud.githubusercontent.com/assets/43060/4467137/a30ee70a-48ec-11e4-9a94-4b4e3f33dc43.png)

And halved our request time.

I'm not sure what meta-programming exactly octopus is doing, but its doing some class definition inside the request cycle in rails 4, which is resetting the constant cache sometimes up to 100s of times per request. 

From the traces generated by Busted, I suspect it's something to do with the AR association proxies. Bigger, more complex requests that load lots of dependent models seems to make the invalidations skyrocket.

**Comment by phallstrom** [comment]
Just starting to look at octopus and saw this issue.  Last update was 7 months ago.  Is it still an issue?

**Comment by hschoenburg** [comment]
Hmm yeah is anyone home here? This is such a cool gem. It would be tragic to see it go unmaintained.

**Comment by matthewrudy** [comment]
I should have found this earlier
I just created a project to showcase the performance degradation.

https://github.com/gogovan/octopus-performance

**Comment by matthewrudy** [comment]
For those interested I build my own library for using read-only replicas
https://github.com/gogovan/active_replica

This only covers part of the Octopus functionality,
but if you just want to manually switch to a read-only replica
(eg. with heroku followers or RDS)
then this made a massive difference to our server performance

<img width="844" alt="screen shot 2015-09-30 at 12 09 17 pm" src="https://cloud.githubusercontent.com/assets/8659/10191855/70c5b430-67aa-11e5-9298-6e72041e6df4.png">

**Comment by dpbus** [comment]
Anyone still looking at this? Could a stop gap simply be wrapping this line in a rails version < 4 conditional? https://github.com/tchandy/octopus/blob/master/lib/octopus/association_shard_tracking.rb#L109
At least then it wouldn't be a problem for anyone using rails 4 or higher.

**Comment by nickmarden** [comment]
Hi @dpbus, sorry I took a while to respond to this. I'm going to poke around to see if there is any easy solution to this issue.

**Comment by nickmarden** [comment]
OK, so I spent a day on this, and came up with [a branch](https://github.com/nickmarden/octopus/tree/remove-ar-extends) that solves the "double-select" issue for Rails 3.x without using ActiveRecord Association extensions.

The only problem is, it doesn't actually cause anything to perform better, which makes we wonder whether the metaprogramming was ever really the issue. To wit:

```
Before Octopus
Calculating -------------------------------------
user.posts - 0 posts - user.posts
                        21.713k i/100ms
user.posts - 0 posts - user.posts.map(&:title)
                        15.134k i/100ms
user.posts - 100 posts - user.posts
                        22.007k i/100ms
user.posts - 100 posts - user.posts.map(&:title)
                         1.342k i/100ms
           post.user    15.833k i/100ms
      post.user.name    14.227k i/100ms
-------------------------------------------------
user.posts - 0 posts - user.posts
                        858.471k (± 7.2%) i/s -      4.277M
user.posts - 0 posts - user.posts.map(&:title)
                        387.745k (± 6.0%) i/s -      1.937M
user.posts - 100 posts - user.posts
                        844.850k (± 8.1%) i/s -      4.203M
user.posts - 100 posts - user.posts.map(&:title)
                         14.687k (± 3.9%) i/s -     73.810k
           post.user    458.774k (± 6.7%) i/s -      2.296M
      post.user.name    349.472k (± 5.6%) i/s -      1.750M
After Octopus
Calculating -------------------------------------
user.posts - 0 posts - user.posts
                         4.677k i/100ms
user.posts - 0 posts - user.posts.map(&:title)
                         1.598k i/100ms
user.posts - 100 posts - user.posts
                         4.693k i/100ms
user.posts - 100 posts - user.posts.map(&:title)
                       732.000  i/100ms
           post.user     4.294k i/100ms
      post.user.name     4.063k i/100ms
-------------------------------------------------
user.posts - 0 posts - user.posts
                         63.703k (± 4.4%) i/s -    318.036k
user.posts - 0 posts - user.posts.map(&:title)
                         17.917k (± 3.1%) i/s -     91.086k
user.posts - 100 posts - user.posts
                         64.713k (± 3.1%) i/s -    323.817k
user.posts - 100 posts - user.posts.map(&:title)
                          7.870k (± 4.0%) i/s -     39.528k
           post.user     58.116k (± 4.4%) i/s -    291.992k
      post.user.name     54.871k (± 3.6%) i/s -    276.284k
```

@matthewrudy would it be possible for you to improve your helpful performance tester to also work with older versions of Rails and Octopus? I'd love to `git bisect` my way to a smoking gun, or demonstrate that Octopus has always had crappy performance, or something...but I've already spent a bunch of hours on it and have wound up empty-handed despite (I think) solving the issue that this thread claimed was the root cause.

**Comment by matthewrudy** [comment]
@nickmarden I'm not exactly sure what the cause is,
but the issue became clear when I was profiling my app and saw this all over the place.

<img width="489" alt="screenshot 2015-10-13 00 32 38" src="https://cloud.githubusercontent.com/assets/8659/10433110/2e42a9f0-7142-11e5-8fd9-b9d4f4882a9a.png">

**Comment by nickmarden** [comment]
@matthewrudy by "this" do you mean the Octopus code that ensures that queries are made on the correct shard? That's intentional. It seems to cost (1.89% + 1.54% + 1.50% + 1.17% =) 6.1% of the total CPU time in the profile you've captured? That's a cost, but it's not a 10x sort of cost like the overall numbers you're showing. Or maybe I just don't know what I'm looking at - I don't recognize the output format.

I'm trying to get back to why many people in this thread said that constant cache invalidation was the reason for poor performance. Do we actually know that is the cause, or is it just speculation because poor performance was accompanied by an uptick in constant cache invalidation?

**Comment by matthewrudy** [comment]
@nickmarden there was no queries to be made,
the association was already loaded via an `includes`

**Comment by nickmarden** [comment]
Ok, I'm not sure how I was supposed to know that from the screencap, but I believe you. And yes, it's true, inserting shard management into AR means that you have overhead. AR methods don't know if they are going to operate from memory or from the DB until late in the resolution process, and the octopus code operates far above the layer where those decisions happen. So there will be times when shard safety is enforced even though no DB access is going to happen during the call. It costs CPU.

I'm trying to get to something I can take action on regarding the original issue reported in this thread. This thread talks extensively about how constant cache invalidation is killing everyone's rails app performance. My goal is to see if removing the thing that is triggering constant cache invalidation, fixes things.

Can anyone try my branch to see if they continue the constant cache getting invalidated? If the AR extensions from octopus were causing that before, the problem should be gone in the branch I posted. @dalyons @dpbus @ashanbrown @pbyrne ... Anyone?

**Comment by matthewrudy** [comment]
@nickmarden unfortunately that's from actual profiling results
so I can't show you the rest of the page.

But if you use ruby-prof on your app, and have already optimised against N+1 issues, you'll still see this.

Give it a try.

Something like this in `rails console`

``` ruby
result = RubyProf.profile do
  app.get("/api/v1/your/endpoint.json")
end

def output_profile_result(profile_result, filename)
  path = File.join("script/profiling/results", filename)
  File.open(path, 'w') do |file|
    RubyProf::CallStackPrinter.new(profile_result).print(file, expansion: 1)
  end
end

output_profile_result(result, "myresult.html")
```

**Comment by matthewrudy** [comment]
@nickmarden but yeah
I already spent a lot of time on this
and instead rewrote the bits of octopus I need.

Namely explicit shard specification.
(I looked into how to efficiently do the octopus thing, and switch connections based on write or read, but decided there was no good way, would be better if ActiveRecord said `connection_pool_for(model, type_of_action)` then it'd be trivial

**Comment by chrisblatchley** [comment]
Just a note, i'm not sure what you (@matthewrudy) found in your profiling, but NewRelic is telling me that most of the middleware time is spend in ActiveRecord::QueryCache#call, as seen in the graph below.

![screen shot 2015-10-22 at 5 01 12 pm](https://cloud.githubusercontent.com/assets/121771/10678186/bf2266e2-78de-11e5-992d-a19e38ce69be.png)

[Here's the source for this method](https://github.com/rails/rails/blob/2a7cf24cb7aab28f483a6772b608e2868a9030ba/activerecord/lib/active_record/query_cache.rb#L30)

I'm going to look into why this is happening. For what it's worth, our specific use case is to use sharding only to split our databases across regions, and our database calls are exclusively read-only. I'll look into this tomorrow and see if I find anything helpful.

**Comment by matthewrudy** [comment]
@chrisblatchley yep, agree.

Also I got middleware as contributing more than it's fair share of time.

After I moved to my own custom replacement all the time gets correctly apportioned to activerecord

https://github.com/gogovan/active_replica

**Comment by sixfeetover** [comment]
Just to be clear, the last vertical line in @chrisblatchley's graph is our deploy that switches to using Octopus.

**Comment by nickmarden** [comment]
I'll repeat my request from earlier: since @dalyons' original comment pointed to constant cache busting as the root issue, and since I've created a branch that removes that behavior...can anyone test that branch to see if it actually does solve the problem?

Alternatively, can anyone create a test case that shows the issue clearly and works with versions of Octopus before and after https://github.com/nickmarden/octopus/commit/846785bbdaac74fdc48bf975a123f8b74c54896f so that I can use `git bisect` to see what the root cause is? 

I really do want to help, but I need something concrete to work with apart from graphs of bespoke applications having performance issues.

**Comment by chrisblatchley** [comment]
@nickmarden sorry, I'm not really familiar with enough of the Rails middleware at this moment to be able to make a test case that adequately demonstrates the problem, though I'm continuing to look into it.

[`Octopus::Proxy#enable_query_cache!`](https://github.com/tchandy/octopus/blob/master/lib/octopus/proxy.rb#L285-L288) seems like it could be a candidate for busting the cache. I compared it with what is being called in vanilla Rails and I'm not sure what the purpose of the call is, but I saw it is being called on every request before anything happens. and it does several operations on each shard.

Also, I tried your branch, assuming you meant [this branch](https://github.com/nickmarden/octopus/tree/remove-ar-extends) solving the double-select issue, and saw no performance improvement.

**Comment by nickmarden** [comment]
OK @chrisblatchley thanks for getting back to me. As soon as you have a simple test case that shows the problem I'll be happy to rummage around for a fix.

**Comment by troszok** [comment]
@nickmarden octpus from your branch is really helping to reduce the cache invalidations:

<img width="422" alt="screen shot 2015-10-29 at 22 10 01" src="https://cloud.githubusercontent.com/assets/17826/10832190/305c48cc-7e8a-11e5-8bcb-aebfc140a9c0.png">

**Comment by nickmarden** [comment]
Thanks @troszok for the feedback. I provide support for one app that uses an older version of Octopus, and unfortunately one of the unrelated changes that occurred between the commit it uses and the one that my app uses, breaks CI for that app. I'll try to find time this week to find and fix that issue, and then I'll be able to merge the constant cache invalidation fix to tchandy/octopus master.

Of course, that still leaves unsolved the CPU issue, for which I could *really really* use a test case.

**Comment by dpbus** [comment]
Thanks for the branch, @nickmarden. We were finally able to test it out and we're see the same improvements to constant cache busting.

![screenshot 2016-01-25 12 30 25](https://cloud.githubusercontent.com/assets/170764/12560377/f962310e-c35f-11e5-8c6d-e75ada61f054.png)

I would love to see this get merged into mainline (and tagged/released if possible). I feel like the CPU concern should be opened as a separate issue if people are still seeing it. Just as another datapoint, we're switching from data_fabric to octopus and see similar performance characteristics between the two.

**Comment by paulczajka** [comment]
@nickmarden I agree with @dpbus : I would love to see the cache invalidation fix merged and released on the mainline.  I feel like that would reduce noise while investigating performance concerns.

**Comment by nickmarden** [comment]
@paulczajka @dpbus Understood. Unfortunately specs that pass in my branch don't pass when I rebase my branch against the current master. It'll take some time to sort out why that is.

**Comment by paulczajka** [comment]
@nickmarden: It looks like pull request #348 for Issue #347  caused the breaks in your specs.  The following patch appears to be one way of resolving the issue on your branch.  Basically this isolates the fix from the above pull request to only operate on block-supplying methods.

```
diff --git a/lib/octopus/relation_proxy.rb b/lib/octopus/relation_proxy.rb
index 0d2ac11..f6a6d38 100644
--- a/lib/octopus/relation_proxy.rb
+++ b/lib/octopus/relation_proxy.rb
@@ -17,7 +17,11 @@ module Octopus
     end
 
     def method_missing(method, *args, &block)
-      @ar_relation.public_send(method, *args, &block)
+      if block
+        @ar_relation.public_send(method, *args, &block)
+      else
+        run_on_shard { @ar_relation.public_send(method, *args, &block) }
+      end
     end
 
     def ==(other)
```

**Comment by nickmarden** [comment]
That looks right @paulczajka ... or at least right-er. I'm still not sure I understand why _all_ blocks need the naked (non-sharded) approach since #348 seems to be solving a problem specifically for the `.using` case. But at any rate, my changes now pass rspec so I'll open a PR and see if I can get this constant cache fix into master. Thanks!

**Comment by anhkind** [comment]
@nickmarden I can see that your last commit https://github.com/nickmarden/octopus/commit/c6dcdf69b9185761cc26a3c8af25e6542623db54 from your branch above (for #352) would fix the current specs from #347, should we merge that in first?

**Comment by nickmarden** [comment]
@anhkind Good idea, see https://github.com/thiagopradi/octopus/pull/356

**Comment by anhkind** [comment]
@nickmarden :+1:  
I can see that your PR #357 passed all the specs, which is awesome!

**Comment by nickmarden** [comment]
@anhkind @jcoleman @paulczajka and everyone else - these fixes have been merged to master.

**Comment by anhkind** [comment]
It's great. Thanks for the hard work @nickmarden :+1:
